File: 2020/2020W27/PreppindataW27.py
import pandas as pd

xlsx = pd.ExcelFile("F:/Data/Input.xlsx")

ppr = pd.read_excel(xlsx,'Preppers')   
loc = pd.read_excel(xlsx,'Location')
info = pd.read_excel(xlsx,'Information',skiprows=1).dropna(subset=['Rating'])

rel_val = {'Rarely Breaks':1, 'Inconsistent':2, 'Fairly Inconsistent':3, 'Fairly Consistent':4, 'Very Consistent':5}

loc['Site'] = loc['Site'].apply(lambda x: x.split('-')[0].strip())
loc['Dummy'] = 0
ppr['Dummy'] = 0

final = ppr.merge(loc, on='Dummy')
final['Check'] = final.apply(lambda x: len(set(map(str.strip,x['Season'].split(','))) & set(map(str.strip,x['Surf Season'].split(',')))), axis=1)
final = final[final['Check'] > 0].drop(['Dummy','Season'],axis=1).copy()

final = final.merge(info, left_on='Site', right_on='Surf Site')
final['Check'] = final.apply(lambda x: len(set(map(str.strip,x['Board Type'].split(','))) & set(map(str.strip,x['Boards'].split(',')))) *
                             len(set(map(str.strip,x['Skill'].split(','))) & set(map(str.strip,x['Skill Level'].split(',')))) , axis=1)
final = final[final['Check'] > 0].drop(['Check','Skill','Board Type','Site'],axis=1).copy()

final['RatingRel'] = final.apply(lambda x: x['Rating'] + rel_val[x['Reliability']]/100, axis=1)
final = final.loc[final.groupby('Name')['RatingRel'].idxmax()][['Name', 'Surf Site', 'Swell Direction', 'Reliability', 'Wind Direction', 
                                                                'Type', 'Boards', 'Skill Level', 'Surf Season','Rating']].copy()

File: 2020/2020W27/preppindataw27.R
library(dplyr)
library(readxl)
library(tidyr)
library(stringr)
library(splitstackshape)

xlsx <- "F:/Data/Input.xlsx"
rel_val <- c('Rarely Breaks', 'Inconsistent', 'Fairly Inconsistent', 'Fairly Consistent', 'Very Consistent')

final <- read_excel(xlsx, sheet = 'Preppers') %>%
  cSplit(., 'Season', ',') %>%
  pivot_longer(., cols=starts_with('Season'), names_to='To Drop', values_to  = 'Season Split', values_drop_na = TRUE) %>%
  select(-'To Drop') %>%
  cSplit(., 'Board Type', ',') %>%
  pivot_longer(., cols=starts_with('Board Type'), names_to='To Drop', values_to  = 'Boards Split', values_drop_na = TRUE) %>%
  select(-'To Drop') %>%
  merge(read_excel(xlsx, sheet = 'Location'),all = TRUE) %>%
  rowwise() %>%
  filter(`Season Split` %in% str_trim(str_split(`Surf Season`,',')[[1]])) %>%
  separate(`Site`,c('Surf Site','To Drop'), sep='\\s-') %>%
  merge(read_excel(xlsx, sheet = 'Information', skip=1) %>% drop_na(), by='Surf Site') %>%
  rowwise() %>%
  filter((`Boards Split` %in% str_trim(str_split(`Boards`,',')[[1]])) & (`Skill` %in% str_trim(str_split(`Skill Level`,',')[[1]]))) %>%
  ungroup() %>%
  select(c('Name', 'Surf Site', 'Swell Direction', 'Reliability', 'Wind Direction', 'Type', 'Boards', 'Skill Level', 'Surf Season','Rating')) %>%
  distinct() %>%
  mutate('RatingRel'= `Rating`+match(`Reliability`,rel_val)/100) %>%
  group_by(`Name`) %>%
  filter(`RatingRel` == max(`RatingRel`)) %>%
  select(-'RatingRel')

View(final)